### PR TITLE
Search window's position settings

### PR DESCRIPTION
**Pull request by onesounds**
image

- Remember Last Position (Default)
- Mouse Screen Center (same old thing)
- Mouse Screen Center Top (New)
- Mouse Screen Left Top
- Mouse Screen Right Top
- Position Reset on Tray (Turning off any one while using multiple monitors causes a location problem.)

**Features**
- Change RememberLastPosition(bool) to LauncherPosition(string)
- It is difficult to respond to screen rotation, so I think the function of always positioning it at the top will help.

**Test Case**
- When you move the window outside the screen and press Reset Position, the window must be visible to the user.
- Each function shall be aligned in a certain position regardless of the monitor resolution.

**What to do later**
- The ability to check the number of monitors and then open them at a specific monitor location. (ex> Monior 2 Screen Center)

### CODE REVIEW

**Comment by onesounds** [comment]
This pr should merge after clock PR merged.

File: Flow.Launcher.Infrastructure/UserSettings/Settings.cs [hunk 1]
```diff
@@ -199,7 +199,7 @@ public bool HideNotifyIcon
         }
         public bool LeaveCmdOpen { get; set; }
         public bool HideWhenDeactive { get; set; } = true;
-        public bool RememberLastLaunchLocation { get; set; }
+        public string LauncherPosition { get; set; } = "RememberLastLaunchLocation";
```
Comment (taooceros) [inline Flow.Launcher.Infrastructure/UserSettings/Settings.cs:202]: I think you have defined the enum? Why not use them?


```suggestion
        public LauncherPositions LauncherPosition { get; set; } = LauncherPositions.RememberLastLaunchLocation;
```

File: Flow.Launcher/MainWindow.xaml.cs [hunk 2]
```diff
@@ -182,25 +182,41 @@ private void OnLoaded(object sender, RoutedEventArgs _)
 
         private void InitializePosition()
         {
-            if (_settings.RememberLastLaunchLocation)
+            if (_settings.LauncherPosition == "RememberLastLaunchLocation")
```
Comment (taooceros) [inline Flow.Launcher/MainWindow.xaml.cs:185]: Should use a switch for this

File: Flow.Launcher/MainWindow.xaml.cs [hunk 3]
```diff
@@ -418,23 +446,38 @@ private void UpdatePosition()
             if (_animating)
                 return;
 
-            if (_settings.RememberLastLaunchLocation)
+            if (_settings.LauncherPosition == "RememberLastLaunchLocation")
```
Comment (taooceros) [inline Flow.Launcher/MainWindow.xaml.cs:449]: same switch for this

File: Flow.Launcher/MainWindow.xaml.cs [hunk 4]
```diff
@@ -472,6 +515,15 @@ public double WindowTop()
             return top;
         }
 
+        public double WindowRight()
```
Comment (taooceros) [inline Flow.Launcher/MainWindow.xaml.cs:518]: what's this for?

File: Flow.Launcher/ViewModel/SettingWindowViewModel.cs [hunk 5]
```diff
@@ -382,6 +382,33 @@ public List<ColorScheme> ColorSchemes
...
 
+
+
+        public class LauncherPosition
+        {
+            public string Display { get; set; }
+            public LauncherPositions Value { get; set; }
+        }
+
+        public List<LauncherPosition> LauncherPositions
```
Comment (taooceros) [inline Flow.Launcher/ViewModel/SettingWindowViewModel.cs:393]: This is clever, we may use this structure for all the combobox. Probably we want to cache this result? But that may break some behavior when changing language.

File: Flow.Launcher/MainWindow.xaml.cs (hunk 4 above)
Comment (onesounds) [reply Flow.Launcher/MainWindow.xaml.cs:518]: for Right (screen) position. it using in top-right option.

File: Flow.Launcher/ViewModel/SettingWindowViewModel.cs (hunk 5 above)
Comment (onesounds) [reply Flow.Launcher/ViewModel/SettingWindowViewModel.cs:393]: I just copied and pasted the existing structure. 😋

File: Flow.Launcher.Infrastructure/UserSettings/Settings.cs (hunk 1 above)
Comment (onesounds) [reply Flow.Launcher.Infrastructure/UserSettings/Settings.cs:202]: applied it.

File: Flow.Launcher/MainWindow.xaml.cs (hunk 3 above)
Comment (onesounds) [reply Flow.Launcher/MainWindow.xaml.cs]: I changed the part to call InitializePosition().

File: Flow.Launcher/MainWindow.xaml.cs (hunk 2 above)
Comment (onesounds) [reply Flow.Launcher/MainWindow.xaml.cs]: Applied it.

File: Flow.Launcher/MainWindow.xaml.cs [hunk 6]
```diff
@@ -182,15 +182,28 @@ private void OnLoaded(object sender, RoutedEventArgs _)
...
+                case LauncherPositions.MouseScreenCenterTop:
+                    Left = WindowLeft();
+                    Top = 10;
+                    break;
+                case LauncherPositions.MouseScreenLeftTop:
+                    Left = 10;
+                    Top = 10;
+                    break;
+                case LauncherPositions.MouseScreenRightTop:
+                    Left = WindowRight();
```
Comment (taooceros) [inline Flow.Launcher/MainWindow.xaml.cs:204]: We may want to change name of all three method here.

**Review by taooceros — DISMISSED** [review]

File: Flow.Launcher/SettingWindow.xaml.cs [hunk 7]
```diff
@@ -435,6 +435,12 @@ private void OnPluginSettingKeydown(object sender, KeyEventArgs e)
                 pluginFilterTxb.Focus();
         }
 
+        private void LauncherPositionSelectedIndexChanged(object sender, SelectionChangedEventArgs e)
```
Comment (taooceros) [inline Flow.Launcher/SettingWindow.xaml.cs:438]: what's this method for?

Comment (onesounds) [reply Flow.Launcher/SettingWindow.xaml.cs:438]: I don't know either. It seems like a code that VS automatically makes. When I'm making something, I sometimes see this blank code.

**Comment by jjw24** [comment]
@taooceros @onesounds could we merge dev in please.

**Comment by onesounds** [comment]
- I confirmed the positioning problem from QuickSizeAdjust feature. (It touched positioning when Adjust Size)
- Fixed conflicts and issues
- Position Reset has small wrong positioning. (Positioning timing issue.)
- I'll fix this part more and merge it.

**Comment by jjw24** [comment]
@onesounds let me know when pr is good to go again.

**Comment by jjw24** [comment]
One thing from my testing prior to your recent commits was that on a fresh install, the query window starts up on the top left corner. I wonder if we should have `Mouse focused screen center` as a default instead?

**Comment by onesounds** [comment]
> One thing from my testing prior to your recent merge Dev commits @onesounds, was that on a fresh install, the query window starts up on the top left corner. I wonder if we should have `Mouse focused screen center` as a default instead?

OK. I noticed little bugs. I'll mention to you when I think it's OK to review

**Comment by jjw24** [comment]
> I think remember last position should default and it should be center when fresh install

This will have the query window stay on the same monitor screen when dual or more monitors are in use, wouldn't the mouse focus option be better for people who use multiple monitors (especially if they are new to flow and don't know you can change those options yet)?

**Comment by onesounds** [comment]
A state that can be placed in the desired position should be the default.

> > I think remember last position should default and it should be center when fresh install
> 
> This will have the query window stay on the same monitor screen when dual or more monitors are in use, wouldn't the mouse focus option be better for people who use multiple monitors (especially if they are new to flow and don't know you can change those options yet)?

Well... OK. I think you are right too. default will be 'mouse focused screen center'.

File: Flow.Launcher/MainWindow.xaml.cs [hunk 8]
```diff
@@ -275,10 +317,17 @@ private void ToggleGameMode()
                 _viewModel.GameModeStatus = true;
             }
         }
+        private async void PositionReset()
+        {
+           _viewModel.Show();
+           await Task.Delay(300); // If don't give a time, Positioning will be weired.
```
Comment (jjw24) [inline Flow.Launcher/MainWindow.xaml.cs:323]:
```suggestion
           await Task.Delay(300); // If don't give a time, Positioning will be weird.
```

**Review by jjw24 — APPROVED** [review]
👍 👍 👍

**Pull request merged**